src/dbnode/client/circuitbreaker/middleware/middleware.go

```diff
@@ -0,0 +1,245 @@
+package middleware
+
+import (
+	"github.com/m3db/m3/src/dbnode/client/circuitbreaker/circuitbreakererror"
+	"github.com/m3db/m3/src/dbnode/client/circuitbreaker/internal/circuitbreaker"
+	"github.com/m3db/m3/src/dbnode/generated/thrift/rpc"
+	"github.com/uber-go/tally"
+	tchannel "github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+	"go.uber.org/zap"
+	"sync/atomic"
+)
+
+// circuitBreakerClient is a client that wraps a TChannel client with a circuit breaker.
+type circuitBreakerClient struct {
```

Comment:
nit: can just call this `type client struct` (it's in the circuitbreaker package, kinda clear what it is for)